optuna/storages/cached_storage.py

```diff
@@ -0,0 +1,291 @@
+
+    def get_study_name_from_id(self, study_id: int) -> str:
+
+        return self._storage.get_study_name_from_id(study_id)
+
+    def get_study_direction(self, study_id: int) -> StudyDirection:
+
+        return self._storage.get_study_direction(study_id)
+
+    def get_study_user_attrs(self, study_id: int) -> Dict[str, Any]:
+
+        return self._storage.get_study_user_attrs(study_id)
+
+    def get_study_system_attrs(self, study_id: int) -> Dict[str, Any]:
+
+        return self._storage.get_study_system_attrs(study_id)
+
+    def get_all_study_summaries(self) -> List[StudySummary]:
+
+        return self._storage.get_all_study_summaries()
+
+    def create_new_trial(self, study_id: int, template_trial: Optional[FrozenTrial] = None) -> int:
+
+        trial_id, frozen_trial = self._storage._create_new_trial_with_trial(
+            study_id, template_trial
+        )
+        with self._lock:
+            self._study_id_to_trial_id[study_id].append(trial_id)
+            self._trial_id_to_study_id[trial_id] = study_id
+            if frozen_trial.state == TrialState.RUNNING:
+                self._hot_cache[trial_id] = frozen_trial, _TrialUpdates()
+            return trial_id
+
+    def set_trial_state(self, trial_id: int, state: TrialState) -> bool:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                cached_trial.state = state
+                updates.state = state
+                if trial_id in self._dirty_trials:
+                    self._dirty_trials.remove(trial_id)
+                ret = self._update_storage(trial_id)
+                if cached_trial.state != TrialState.RUNNING:
+                    del self._hot_cache[trial_id]
+                return ret
+
+        return self._storage.set_trial_state(trial_id, state)
+
+    def set_trial_param(
+        self,
+        trial_id: int,
+        param_name: str,
+        param_value_internal: float,
+        distribution: distributions.BaseDistribution,
+    ) -> bool:
+
+        with self._lock:
+            if trial_id in self._hot_cache:
+                cached_trial, updates = self._hot_cache[trial_id]
+                study_id = self._trial_id_to_study_id[trial_id]
+                cached_dist = self._param_distributions[study_id].get(param_name, None)
+                if cached_dist:
+                    distributions.check_distribution_compatibility(cached_dist, distribution)
+                else:
+                    self._param_distributions[study_id][param_name] = distribution
+                if param_name in cached_trial.params:
+                    return False
+                self._dirty_trials.add(trial_id)
+                cached_trial.params[param_name] = distribution.to_external_repr(
+                    param_value_internal
+                )
+                cached_trial.distributions[param_name] = distribution
+                if cached_dist:
```

Comment:
I guess the line

```
self._dirty_trials.add(trial_id)
```

should be moved under `if cached_dist:` .

The logic here is a bit difficult too. Some comments, like "If cached_dist isn't available, we have to access the storage in order to check distribution compatibility." would help readers.

Reply:
Thank you for your comment. Indeed, the line should be moved, and the whole logic requires more readability. I'll address them.